### Answers to Assignment 6 - Pandas Fundamentals

Complete the tasks below. Please turn in a single Jupyter notebook named `6_first_last.ipynb` (substitute your first and last name). Please run Kernel > Restart & Run All on your notebook before turning in.

#### The Earth Microbiome Project

For this assignment, we will use Pandas to examine metadata from the [Earth Microbiome Project](http://earthmicrobiome.org/).

First, download the metadata file for a 2,000-sample subset of the >27,000 samples in the Release 1 16S rRNA dataset.

```
curl -O "ftp://ftp.microbio.me/emp/release1/mapping_files/emp_qiime_mapping_subset_2k.tsv"
```

In [1]:
# import the required pacakges
import pandas as pd
import numpy as np

In [2]:
# set the maximum number of rows displayed
pd.set_option("display.max_rows", 10)

#### Answer to A. Reading and summarizing

Q1. Import the tab-separated values file `emp_qiime_mapping_subset_2k.tsv` as a DataFrame called `df` with default data types, with the first row as column labels (columns) and the first column as row labels (indexes).

In [3]:
df = pd.read_csv('../../data/emp_qiime_mapping_subset_2k.tsv', sep='\t', index_col=0)

Q2. The indexes should be the sample IDs. How many samples are in this DataFrame? How many metadata columns?

In [4]:
df.shape

(2000, 75)

Q3. What are the minimum and maximum pH values in the dataset?

In [5]:
df.ph.min(), df.ph.max()

(3.45, 12.3)

Q4. What are the average and standard deviation temperature values in the dataset?

In [6]:
df.describe()

study_id  read_length_bp  sequences_split_libraries  \
count  2000.000000     2000.000000                2000.000000   
mean   1423.491000      125.751000              118683.372000   
std     540.208383       24.318937              100946.740563   
min     550.000000       90.000000               10470.000000   
25%     925.000000      100.000000               53760.500000   
50%    1521.000000      138.000000               92211.500000   
75%    1773.000000      150.000000              148746.750000   
max    2382.000000      151.000000              819180.000000   

       observations_closed_ref_greengenes  observations_closed_ref_silva  \
count                         2000.000000                    2000.000000   
mean                         91485.516000                   94659.511500   
std                          85329.714404                   87431.980992   
min                          10030.000000                   10253.000000   
25%                          36146.000000                   37562.500000   
50%                          66816.000000                   69017.000000   
75%                         117519.250000                  120490.500000   
max                         713117.000000                  717450.000000   

       observations_open_ref_greengenes  observations_deblur_90bp  \
count                       2000.000000               2000.000000   
mean                      115786.274500              64789.053500   
std                        99960.933222              58928.224713   
min                        10457.000000               5275.000000   
25%                        49896.500000              28138.750000   
50%                        89555.500000              50559.500000   
75%                       144549.750000              81776.750000   
max                       814606.000000             589438.000000   

       observations_deblur_100bp  observations_deblur_150bp  sample_taxid  \
count                2000.000000                2000.000000  2.000000e+03   
mean                58213.264000               18119.118000  6.451546e+05   
std                 55653.344075               30223.641917  2.715913e+05   
min                     0.000000                   0.000000  5.549700e+04   
25%                 22395.750000                   0.000000  4.122310e+05   
50%                 47044.500000                4042.000000  5.561820e+05   
75%                 74205.000000               26975.000000  7.499060e+05   
max                573024.000000              294681.000000  1.649191e+06   

              ...          adiv_shannon  adiv_faith_pd  temperature_deg_c  \
count         ...           2000.000000    2000.000000         411.000000   
mean          ...              5.336688      45.323499          18.491088   
std           ...              2.540171      42.088118          16.672381   
min           ...             -0.000000       1.432725         -15.000000   
25%           ...              3.728895      14.737926           9.714750   
50%           ...              5.544158      30.635928          15.240000   
75%           ...              7.202397      61.773213          20.207798   
max           ...             10.738627     257.719293          91.000000   

               ph  salinity_psu  oxygen_mg_per_l  phosphate_umol_per_l  \
count  284.000000    121.000000        26.000000            144.000000   
mean     8.123576     24.959681        10.935769             14.278391   
std      1.684302     12.954485         1.676604             36.338796   
min      3.450000      0.043900         8.400000              0.002527   
25%      7.300000      7.440000         9.300000              0.348971   
50%      8.061450     31.400000        10.650000              0.955000   
75%      9.000000     34.966800        12.137500              4.090000   
max     12.300000     37.065900        13.800000            205.000000   

       ammonium_umol_per_l  nitrate_umol_per_l  sulfate_umol_p

**Answers:**

* There are 2000 samples and 75 metadata columns.
* The min and max pH values are 3.45, 12.3.
* The mean and std temperature values are 18.5, 16.7.

#### Answer to B. Indexing, slicing, and writing

Q1. Make a new Series called `temp` with the temperature column as its own Series object. Remove NaN values (`np.nan`) from this Series. How many values are left?

In [7]:
temp = df.temperature_deg_c

In [8]:
temp.dropna(inplace=True)

In [9]:
temp

#SampleID
678.OA.mesocosm.362                                       11.000000
678.OA.mesocosm.376                                       10.800000
678.OA.mesocosm.410                                       12.100000
678.OA.mesocosm.417                                       11.000000
678.OA.mesocosm.431                                       11.100000
                                                            ...    
2229.W2.T3.4.HP1.Thomas.CMB.Seaweed.lane5.NoIndex.L005    18.300000
2229.W2.T33.PS5.Thomas.CMB.Seaweed.lane6.NoIndex.L006     20.236952
2300.BB.4087.anus                                         35.055556
2300.BB.4087.lavage20                                     35.055556
2300.BB.08.lavage20                                       36.222222
Name: temperature_deg_c, Length: 411, dtype: float64

Q2. Make a new DataFrame called `df_seqs` from columns `sequences_split_libraries` through `observations_deblur_150bp` (column positions 17-23) of the existing DataFrame. What is the mean value of `column observations_deblur_90bp`?

In [10]:
df_seqs = df.iloc[:, 17:24]
df_seqs.head()

sequences_split_libraries  \
#SampleID                                            
550.L1S116.s.1.sequence                      33383   
550.L1S119.s.1.sequence                      40944   
550.L1S164.s.1.sequence                      35636   
550.L1S194.s.1.sequence                      46992   
550.L1S20.s.1.sequence                       30131   

                         observations_closed_ref_greengenes  \
#SampleID                                                     
550.L1S116.s.1.sequence                               32153   
550.L1S119.s.1.sequence                               39472   
550.L1S164.s.1.sequence                               34550   
550.L1S194.s.1.sequence                               43925   
550.L1S20.s.1.sequence                                29179   

                         observations_closed_ref_silva  \
#SampleID                                                
550.L1S116.s.1.sequence                          32453   
550.L1S119.s.1.sequence                          39929   
550.L1S164.s.1.sequence                          34666   
550.L1S194.s.1.sequence                          43852   
550.L1S20.s.1.sequence                           29553   

                         observations_open_ref_greengenes  \
#SampleID                                                   
550.L1S116.s.1.sequence                             33337   
550.L1S119.s.1.sequence                             40870   
550.L1S164.s.1.sequence                             35599   
550.L1S194.s.1.sequence                             46875   
550.L1S20.s.1.sequence                              30094   

                         observations_deblur_90bp  observations_deblur_100bp  \
#SampleID                                                                      
550.L1S116.s.1.sequence                     22567                      22160   
550.L1S119.s.1.sequence                     27871                      27191   
550.L1S164.s.1.sequence                     24134                      23686   
550.L1S194.s.1.sequence                     30041                      29264   
550.L1S20.s.1.sequence                      21132                      20643   

                         observations_deblur_150bp  
#SampleID                                           
550.L1S116.s.1.sequence                       1043  
550.L1S119.s.1.sequence                       1272  
550.L1S164.s.1.sequence                       1161  
550.L1S194.s.1.sequence                       1974  
550.L1S20.s.1.sequence                         603

We can get the means from `describe()`.

In [11]:
df_seqs.describe()

sequences_split_libraries  observations_closed_ref_greengenes  \
count                2000.000000                         2000.000000   
mean               118683.372000                        91485.516000   
std                100946.740563                        85329.714404   
min                 10470.000000                        10030.000000   
25%                 53760.500000                        36146.000000   
50%                 92211.500000                        66816.000000   
75%                148746.750000                       117519.250000   
max                819180.000000                       713117.000000   

       observations_closed_ref_silva  observations_open_ref_greengenes  \
count                    2000.000000                       2000.000000   
mean                    94659.511500                     115786.274500   
std                     87431.980992                      99960.933222   
min                     10253.000000                      10457.000000   
25%                     37562.500000                      49896.500000   
50%                     69017.000000                      89555.500000   
75%                    120490.500000                     144549.750000   
max                    717450.000000                     814606.000000   

       observations_deblur_90bp  observations_deblur_100bp  \
count               2000.000000                2000.000000   
mean               64789.053500               58213.264000   
std                58928.224713               55653.344075   
min                 5275.000000                   0.000000   
25%                28138.750000               22395.750000   
50%                50559.500000               47044.500000   
75%                81776.750000               74205.000000   
max               589438.000000              573024.000000   

       observations_deblur_150bp  
count                2000.000000  
mean                18119.118000  
std                 30223.641917  
min                     0.000000  
25%                     0.000000  
50%                  4042.000000  
75%                 26975.000000  
max                294681.000000

...or we can get the mean on a single column (Series) using `mean()`.

In [12]:
df_seqs.observations_deblur_90bp.mean()

64789.0535

Q3. Save `df_seqs` as a csv file.

In [13]:
df_seqs.to_csv('emp_qiime_mapping_subset_2k_seqs.csv')

**Answers:**

* There are 411 non-NaN temperature values.
* The mean value of observations_deblur_90bp is 64789.0535.

#### Answers to C. Merging, joining, and concatenating

Q1. Store the first 5 rows of `df_seqs` as a new dataframe called `df_seqs_head`. Store the last 5 rows of `df_seqs` as a new dataframe called `df_seqs_tail`.

In [14]:
df_seqs_head = df_seqs.head()

In [15]:
df_seqs_tail = df_seqs.tail()

Q2. Concatenate `df_seqs_head` and `df_seqs_tail` using the `concat()` function.

In [16]:
pd.concat([df_seqs_head, df_seqs_tail])

sequences_split_libraries  \
#SampleID                                                                       
550.L1S116.s.1.sequence                                                 33383   
550.L1S119.s.1.sequence                                                 40944   
550.L1S164.s.1.sequence                                                 35636   
550.L1S194.s.1.sequence                                                 46992   
550.L1S20.s.1.sequence                                                  30131   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                     125047   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                     138753   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                     345657   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                      89747   
2382.DPOO1.C1.HA.1.228.root.9.11.lane1.NoIndex....                      92336   

                                                    observations_closed_ref_greengenes  \
#SampleID                                                                                
550.L1S116.s.1.sequence                                                          32153   
550.L1S119.s.1.sequence                                                          39472   
550.L1S164.s.1.sequence                                                          34550   
550.L1S194.s.1.sequence                                                          43925   
550.L1S20.s.1.sequence                                                           29179   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                              124344   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                              138268   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                              243403   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                               62332   
2382.DPOO1.C1.HA.1.228.root.9.11.lane1.NoIndex....                               66419   

                                                    observations_closed_ref_silva  \
#SampleID                                                                           
550.L1S116.s.1.sequence                                                     32453   
550.L1S119.s.1.sequence                                                     39929   
550.L1S164.s.1.sequence                                                     34666   
550.L1S194.s.1.sequence                                                     43852   
550.L1S20.s.1.sequence                                                      29553   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                         123968   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                         138165   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                         272832   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                          70308   
2382.DPOO1.C1.HA.1.228.root.9.11.lane1.NoIndex....                          71897   

                                                    observations_open_ref_greengenes  \
#SampleID                                                                              
550.L1S116.s.1.sequence                                                        33337   
550.L1S119.s.1.sequence                                                        40870   
550.L1S164.s.1.sequence                                                        35599   
550.L1S194.s.1.sequence                                                        46875   
550.L1S20.s.1.sequence                                                         30094   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                            124983   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                            138702   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                            342881   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                             88986   
2382.DPOO

Q3. Append `df_seqs_tail` to `df_seqs_head` using the `append()` function.

In [17]:
df_seqs_head.append(df_seqs_tail)

sequences_split_libraries  \
#SampleID                                                                       
550.L1S116.s.1.sequence                                                 33383   
550.L1S119.s.1.sequence                                                 40944   
550.L1S164.s.1.sequence                                                 35636   
550.L1S194.s.1.sequence                                                 46992   
550.L1S20.s.1.sequence                                                  30131   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                     125047   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                     138753   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                     345657   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                      89747   
2382.DPOO1.C1.HA.1.228.root.9.11.lane1.NoIndex....                      92336   

                                                    observations_closed_ref_greengenes  \
#SampleID                                                                                
550.L1S116.s.1.sequence                                                          32153   
550.L1S119.s.1.sequence                                                          39472   
550.L1S164.s.1.sequence                                                          34550   
550.L1S194.s.1.sequence                                                          43925   
550.L1S20.s.1.sequence                                                           29179   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                              124344   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                              138268   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                              243403   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                               62332   
2382.DPOO1.C1.HA.1.228.root.9.11.lane1.NoIndex....                               66419   

                                                    observations_closed_ref_silva  \
#SampleID                                                                           
550.L1S116.s.1.sequence                                                     32453   
550.L1S119.s.1.sequence                                                     39929   
550.L1S164.s.1.sequence                                                     34666   
550.L1S194.s.1.sequence                                                     43852   
550.L1S20.s.1.sequence                                                      29553   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                         123968   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                         138165   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                         272832   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                          70308   
2382.DPOO1.C1.HA.1.228.root.9.11.lane1.NoIndex....                          71897   

                                                    observations_open_ref_greengenes  \
#SampleID                                                                              
550.L1S116.s.1.sequence                                                        33337   
550.L1S119.s.1.sequence                                                        40870   
550.L1S164.s.1.sequence                                                        35599   
550.L1S194.s.1.sequence                                                        46875   
550.L1S20.s.1.sequence                                                         30094   
2382.DPOO1.C1.HA.1.630.gp.9.12.lane8.NoIndex.L0...                            124983   
2382.DPOO1.C1.HA.1.629.leav.9.12.lane8.NoIndex....                            138702   
2382.DPOO1.C1.HA.1.628.root.9.12.lane7.NoIndex....                            342881   
2382.DPOO1.C1.HA.1.428.root.4.12.lane7.NoIndex....                             88986   
2382.DPOO

Q4. Make a new DataFrame called `df_phys` with the pH, temperature, and salinity columns (hint: you will need to know the exact column names; these are some of the last few columns). Make another new DataFrame called `df_empo` with the column `empo_3` (note: this will actually be a Series because it has only one column, but you can treat it like a DataFrame).

In [18]:
df_phys = df[['ph', 'temperature_deg_c', 'salinity_psu']]

In [19]:
df_empo = df['empo_3']

Q5. Merge `df_phys` with `df_seqs` using the `merge()` function with the indexes of both DataFrames to make a new DataFrame called `df_merged`.

In [20]:
df_merged = pd.merge(df_phys, df_seqs, left_index=True, right_index=True)
df_merged.head()

ph  temperature_deg_c  salinity_psu  \
#SampleID                                                      
550.L1S116.s.1.sequence NaN                NaN           NaN   
550.L1S119.s.1.sequence NaN                NaN           NaN   
550.L1S164.s.1.sequence NaN                NaN           NaN   
550.L1S194.s.1.sequence NaN                NaN           NaN   
550.L1S20.s.1.sequence  NaN                NaN           NaN   

                         sequences_split_libraries  \
#SampleID                                            
550.L1S116.s.1.sequence                      33383   
550.L1S119.s.1.sequence                      40944   
550.L1S164.s.1.sequence                      35636   
550.L1S194.s.1.sequence                      46992   
550.L1S20.s.1.sequence                       30131   

                         observations_closed_ref_greengenes  \
#SampleID                                                     
550.L1S116.s.1.sequence                               32153   
550.L1S119.s.1.sequence                               39472   
550.L1S164.s.1.sequence                               34550   
550.L1S194.s.1.sequence                               43925   
550.L1S20.s.1.sequence                                29179   

                         observations_closed_ref_silva  \
#SampleID                                                
550.L1S116.s.1.sequence                          32453   
550.L1S119.s.1.sequence                          39929   
550.L1S164.s.1.sequence                          34666   
550.L1S194.s.1.sequence                          43852   
550.L1S20.s.1.sequence                           29553   

                         observations_open_ref_greengenes  \
#SampleID                                                   
550.L1S116.s.1.sequence                             33337   
550.L1S119.s.1.sequence                             40870   
550.L1S164.s.1.sequence                             35599   
550.L1S194.s.1.sequence                             46875   
550.L1S20.s.1.sequence                              30094   

                         observations_deblur_90bp  observations_deblur_100bp  \
#SampleID                                                                      
550.L1S116.s.1.sequence                     22567                      22160   
550.L1S119.s.1.sequence                     27871                      27191   
550.L1S164.s.1.sequence                     24134                      23686   
550.L1S194.s.1.sequence                     30041                      29264   
550.L1S20.s.1.sequence                      21132                      20643   

                         observations_deblur_150bp  
#SampleID                                           
550.L1S116.s.1.sequence                       1043  
550.L1S119.s.1.sequence                       1272  
550.L1S164.s.1.sequence                       1161  
550.L1S194.s.1.sequence                       1974  
550.L1S20.s.1.sequence                         603

Q6. Join `df_merged` with `df_empo` using the `join()` function and store the result as `df_merged`.

In [21]:
df_merged = df_merged.join(df_empo)
df_merged.head()

ph  temperature_deg_c  salinity_psu  \
#SampleID                                                      
550.L1S116.s.1.sequence NaN                NaN           NaN   
550.L1S119.s.1.sequence NaN                NaN           NaN   
550.L1S164.s.1.sequence NaN                NaN           NaN   
550.L1S194.s.1.sequence NaN                NaN           NaN   
550.L1S20.s.1.sequence  NaN                NaN           NaN   

                         sequences_split_libraries  \
#SampleID                                            
550.L1S116.s.1.sequence                      33383   
550.L1S119.s.1.sequence                      40944   
550.L1S164.s.1.sequence                      35636   
550.L1S194.s.1.sequence                      46992   
550.L1S20.s.1.sequence                       30131   

                         observations_closed_ref_greengenes  \
#SampleID                                                     
550.L1S116.s.1.sequence                               32153   
550.L1S119.s.1.sequence                               39472   
550.L1S164.s.1.sequence                               34550   
550.L1S194.s.1.sequence                               43925   
550.L1S20.s.1.sequence                                29179   

                         observations_closed_ref_silva  \
#SampleID                                                
550.L1S116.s.1.sequence                          32453   
550.L1S119.s.1.sequence                          39929   
550.L1S164.s.1.sequence                          34666   
550.L1S194.s.1.sequence                          43852   
550.L1S20.s.1.sequence                           29553   

                         observations_open_ref_greengenes  \
#SampleID                                                   
550.L1S116.s.1.sequence                             33337   
550.L1S119.s.1.sequence                             40870   
550.L1S164.s.1.sequence                             35599   
550.L1S194.s.1.sequence                             46875   
550.L1S20.s.1.sequence                              30094   

                         observations_deblur_90bp  observations_deblur_100bp  \
#SampleID                                                                      
550.L1S116.s.1.sequence                     22567                      22160   
550.L1S119.s.1.sequence                     27871                      27191   
550.L1S164.s.1.sequence                     24134                      23686   
550.L1S194.s.1.sequence                     30041                      29264   
550.L1S20.s.1.sequence                      21132                      20643   

                         observations_deblur_150bp             empo_3  
#SampleID                                                              
550.L1S116.s.1.sequence                       1043  Animal distal gut  
550.L1S119.s.1.sequence                       1272  Animal distal gut  
550.L1S164.s.1.sequence                       1161  Animal distal gut  
550.L1S194.s.1.sequence                       1974  Animal distal gut  
550.L1S20.s.1.sequence                         603  Animal distal gut

#### Answer to D. Applying functions

Q1. Use a list comprehension to add a new column to `df_merged` called `temperature_deg_f` which takes the values in `temperature_deg_c` and converts them to degrees Fahrenheit.

In [22]:
df_merged['temperature_deg_f'] = [c * 1.8 + 32 for c in df_merged.temperature_deg_c]

Q2. Create a function that changes a single numerical value from Celsius to Fahrenheit. Apply this function to the values in `temperature_deg_c` using `apply()` to create a new column called `temperature_deg_f_2`.

In [23]:
def celsius_to_fahrenheit(c):
    f = c * 1.8 + 32
    return(f)

In [24]:
df_merged['temperature_deg_f_2'] = df_merged['temperature_deg_c'].apply(celsius_to_fahrenheit)

To see the results, we can select the rows that have non-NA temperature values (not all samples have temperature data) and the columns containing the temperature data.

In [25]:
df_merged.loc[df_merged.temperature_deg_c.notna(), 
              ['temperature_deg_c', 'temperature_deg_f', 'temperature_deg_f_2']]

temperature_deg_c  \
#SampleID                                                               
678.OA.mesocosm.362                                         11.000000   
678.OA.mesocosm.376                                         10.800000   
678.OA.mesocosm.410                                         12.100000   
678.OA.mesocosm.417                                         11.000000   
678.OA.mesocosm.431                                         11.100000   
...                                                               ...   
2229.W2.T3.4.HP1.Thomas.CMB.Seaweed.lane5.NoInd...          18.300000   
2229.W2.T33.PS5.Thomas.CMB.Seaweed.lane6.NoInde...          20.236952   
2300.BB.4087.anus                                           35.055556   
2300.BB.4087.lavage20                                       35.055556   
2300.BB.08.lavage20                                         36.222222   

                                                    temperature_deg_f  \
#SampleID                                                               
678.OA.mesocosm.362                                         51.800000   
678.OA.mesocosm.376                                         51.440000   
678.OA.mesocosm.410                                         53.780000   
678.OA.mesocosm.417                                         51.800000   
678.OA.mesocosm.431                                         51.980000   
...                                                               ...   
2229.W2.T3.4.HP1.Thomas.CMB.Seaweed.lane5.NoInd...          64.940000   
2229.W2.T33.PS5.Thomas.CMB.Seaweed.lane6.NoInde...          68.426513   
2300.BB.4087.anus                                           95.100000   
2300.BB.4087.lavage20                                       95.100000   
2300.BB.08.lavage20                                         97.200000   

                                                    temperature_deg_f_2  
#SampleID                                                                
678.OA.mesocosm.362                                           51.800000  
678.OA.mesocosm.376                                           51.440000  
678.OA.mesocosm.410                                           53.780000  
678.OA.mesocosm.417                                           51.800000  
678.OA.mesocosm.431                                           51.980000  
...                                                                 ...  
2229.W2.T3.4.HP1.Thomas.CMB.Seaweed.lane5.NoInd...            64.940000  
2229.W2.T33.PS5.Thomas.CMB.Seaweed.lane6.NoInde...            68.426513  
2300.BB.4087.anus                                             95.100000  
2300.BB.4087.lavage20                                         95.100000  
2300.BB.08.lavage20                                           97.200000  

[411 rows x 3 columns]

#### Answer to E. Sorting

Q1. Sort the rows in `df_merged` by `sequences_split_libraries` values from high to low and store the result as `df_merged`. (Hint: you can use `inplace=True`.)

In [26]:
df_merged.sort_values(by='sequences_split_libraries', ascending=False, inplace=True)
df_merged.head()

ph  temperature_deg_c  salinity_psu  \
#SampleID                                                  
1240.1512BCDNA      NaN              10.37         34.39   
1222.B5.5.14.06     NaN               9.30         31.50   
810.1230C2H2        NaN                NaN           NaN   
638.FRX7.120910.2   NaN                NaN           NaN   
1242.ME04Jun01EB1R2 NaN                NaN           NaN   

                     sequences_split_libraries  \
#SampleID                                        
1240.1512BCDNA                          819180   
1222.B5.5.14.06                         747257   
810.1230C2H2                            737194   
638.FRX7.120910.2                       716093   
1242.ME04Jun01EB1R2                     699690   

                     observations_closed_ref_greengenes  \
#SampleID                                                 
1240.1512BCDNA                                   657762   
1222.B5.5.14.06                                  709386   
810.1230C2H2                                     713117   
638.FRX7.120910.2                                593535   
1242.ME04Jun01EB1R2                              633858   

                     observations_closed_ref_silva  \
#SampleID                                            
1240.1512BCDNA                              672649   
1222.B5.5.14.06                             713535   
810.1230C2H2                                717450   
638.FRX7.120910.2                           609665   
1242.ME04Jun01EB1R2                         635995   

                     observations_open_ref_greengenes  \
#SampleID                                               
1240.1512BCDNA                                 814606   
1222.B5.5.14.06                                744797   
810.1230C2H2                                   734793   
638.FRX7.120910.2                              705982   
1242.ME04Jun01EB1R2                            696427   

                     observations_deblur_90bp  observations_deblur_100bp  \
#SampleID                                                                  
1240.1512BCDNA                         539745                     545610   
1222.B5.5.14.06                        555429                     550169   
810.1230C2H2                           460415                     396263   
638.FRX7.120910.2                      398992                     358234   
1242.ME04Jun01EB1R2                    433043                     385579   

                     observations_deblur_150bp              empo_3  \
#SampleID                                                            
1240.1512BCDNA                          234531      Water (saline)   
1222.B5.5.14.06                         294681      Water (saline)   
810.1230C2H2                                 0   Sediment (saline)   
638.FRX7.120910.2                            0  Water (non-saline)   
1242.ME04Jun01EB1R2                          0  Water (non-saline)   

                     temperature_deg_f  temperature_deg_f_2  
#SampleID                                                    
1240.1512BCDNA                  50.666               50.666  
1222.B5.5.14.06                 48.740               48.740  
810.1230C2H2                       NaN                  NaN  
638.FRX7.120910.2                  NaN                  NaN  
1242.ME04Jun01EB1R2                NaN                  NaN

Q2. Sort the columns in `df_merged` by column name from A to Z and store the result as `df_merged`. (Hint: you can use `inplace=True`.)

In [27]:
df_merged.sort_index(axis=1, ascending=True, inplace=True)
df_merged.head()

empo_3  observations_closed_ref_greengenes  \
#SampleID                                                                     
1240.1512BCDNA           Water (saline)                              657762   
1222.B5.5.14.06          Water (saline)                              709386   
810.1230C2H2          Sediment (saline)                              713117   
638.FRX7.120910.2    Water (non-saline)                              593535   
1242.ME04Jun01EB1R2  Water (non-saline)                              633858   

                     observations_closed_ref_silva  observations_deblur_100bp  \
#SampleID                                                                       
1240.1512BCDNA                              672649                     545610   
1222.B5.5.14.06                             713535                     550169   
810.1230C2H2                                717450                     396263   
638.FRX7.120910.2                           609665                     358234   
1242.ME04Jun01EB1R2                         635995                     385579   

                     observations_deblur_150bp  observations_deblur_90bp  \
#SampleID                                                                  
1240.1512BCDNA                          234531                    539745   
1222.B5.5.14.06                         294681                    555429   
810.1230C2H2                                 0                    460415   
638.FRX7.120910.2                            0                    398992   
1242.ME04Jun01EB1R2                          0                    433043   

                     observations_open_ref_greengenes  ph  salinity_psu  \
#SampleID                                                                 
1240.1512BCDNA                                 814606 NaN         34.39   
1222.B5.5.14.06                                744797 NaN         31.50   
810.1230C2H2                                   734793 NaN           NaN   
638.FRX7.120910.2                              705982 NaN           NaN   
1242.ME04Jun01EB1R2                            696427 NaN           NaN   

                     sequences_split_libraries  temperature_deg_c  \
#SampleID                                                           
1240.1512BCDNA                          819180              10.37   
1222.B5.5.14.06                         747257               9.30   
810.1230C2H2                            737194                NaN   
638.FRX7.120910.2                       716093                NaN   
1242.ME04Jun01EB1R2                     699690                NaN   

                     temperature_deg_f  temperature_deg_f_2  
#SampleID                                                    
1240.1512BCDNA                  50.666               50.666  
1222.B5.5.14.06                 48.740               48.740  
810.1230C2H2                       NaN                  NaN  
638.FRX7.120910.2                  NaN                  NaN  
1242.ME04Jun01EB1R2                NaN                  NaN